lib/hyacinth/client.rb

```diff
@@ -0,0 +1,254 @@
+          raise Hyacinth::Client::Exceptions::UnexpectedResponse, "Errors encountered during create_resource attempt: #{user_errors.inspect}" if user_errors.present?
+        end
+      end
+
+      true
+    end
+
+    # @return An Active Storage signed id upon successful upload, or nil if the upload fails.
+    def upload_file_to_active_storage(file_path, original_filename)
+      File.open(file_path, 'rb') do |file|
+        # The line below does a buffered file read for low memory usage, and is the same method
+        # used for checksum calculation as ActiveStorage (to ensure matching checksum validation).
+        file_checksum = Digest::MD5.file(file_path).base64digest
+
+        mime_type = BestType.mime_type.for_file_name(file_path)
+        file_size = file.size
+
+        active_storage_blob_creation_response = @conn.post('/api/v1/uploads') do |req|
+          req.params['blob'] = {
+            filename: original_filename,
+            byte_size: file_size,
+            checksum: file_checksum,
+            content_type: mime_type
+          }
+        end
+
+        signed_id, direct_upload_data = JSON.parse(active_storage_blob_creation_response.body).values_at('signed_id', 'direct_upload')
+
+        active_storage_upload_response = @conn.put(direct_upload_data['url']) do |req|
+          # Required for streaming the upload
+          req.headers.merge!('Transfer-Encoding' => 'chunked', 'Content-Length' => file_size.to_s)
+
+          # Direct upload data includes headers that should be sent in upload request (e.g. Content-Type)
+          req.headers.merge!(direct_upload_data['headers'])
+
+          # Actual request body IS the file content
+          req.body = Faraday::FilePart.new(file.path, mime_type)
+        end
+
+        if active_storage_upload_response.status != ACTIVE_STORAGE_UPLOAD_SUCCESS_STATUS
+          raise Hyacinth::Client::Exceptions::UnexpectedResponse, "Expected #{ACTIVE_STORAGE_UPLOAD_SUCCESS_STATUS} status from Hyacinth for upload, but got #{active_storage_upload_response.status}"
+        end
+
+        signed_id
+      end
+    end
+
+    def resource_request_success!(resource_request_id)
+      update_resource_request(resource_request_id: resource_request_id, status: 'success')
+    end
+
+    def resource_request_failure!(resource_request_id, processing_errors)
+      update_resource_request(resource_request_id: resource_request_id, status: 'failure', processing_errors: processing_errors)
+    end
+
+    def resource_request_in_progress!(resource_request_id)
+      update_resource_request(resource_request_id: resource_request_id, status: 'in_progress')
+    end
+
+    def update_resource_request(resource_request_id:, status:, processing_errors: [])
+      query = <<~GQL
+        mutation ($input: UpdateResourceRequestInput!) {
+          updateResourceRequest(input: $input) {
+            resourceRequest {
+              id
+            }
+          }
+        }
+      GQL
+
```

Comment:
I notice there are no userErrors being returned here. Is that intentional?

Reply:
Good question!  The updateResourceRequest mutation doesn't return `userErrors` within the `data` field right now, so it is intentional, but at some point it might be good to update a bunch of the mutations (particularly any of the ones that involving updates to ActiveRecord objects).